File: 5. Top K Frequent Elements (347)/Top K Frequent Elements (347).cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        int len = nums.size();
        vector<vector<int>> a(len);
        unordered_map<int, int> m;
        vector<int> res;

        for(auto x: nums){
            ++m[x];
        }

        for(auto& [key, value]: m){
            a[value-1].push_back(key);
        }

        for(int i=a.size()-1; i>=0; --i){
            if(k && a[i].size()!=0){
                for(auto x: a[i]){
                    res.push_back(x);
                }
                k=k-a[i].size();
            }
        }
        return res;    
    }
};

File: 5. Top K Frequent Elements (347)/Top K Frequent Elements (347).py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = {}
        topK = [[] for i in range(len(nums))]
        res = []

        for n in nums:
            freq[n] = 1+freq.get(n,0)

        for n, count in freq.items():
            topK[count-1].append(n)
        
        for i in range(len(nums)-1, -1, -1):
            if not k:
                break
            k = k - len(topK[i])
            res = res + topK[i]
        
        return res